File: 22_Change/perl/change.pl
#!/usr/bin/perl

use v5.24; # for say and use strict
use warnings;

sub get_pennies {
  my $query = shift;

  print "$query? ";
  my $in = <>;
  chomp $in;
  $in =~ /([\d.]+)/; # the first match of digits and decimal points
  return int( $1 * 100 );
}

sub make_change {
  my $change = shift;

  state %change_options = (
    'Penny' => { value => 1, plural => 'Pennies' },
    'Nickel' => { value => 5 },
    'Dime' => { value => 10 },
    'Quarter' => { value => 25 },
    'One Half Dollar' => { value => 50 },
    'One Dollar Bill' => { value => 100 * 1 },
    'Five Dollar Bill' => { value => 100 * 5 },
    '10 Dollar Bill' => { value => 100 * 10 },
  );

  foreach my $unit ( sort { $change_options{$b}->{value} <=> $change_options{$a}->{value} } keys %change_options ) {
    my $value = $change_options{$unit}->{value};
    next if $value > $change;
    my $number = int( $change / $value );
    if ( $number > 1 ) {
      $unit = exists $change_options{$unit}->{plural} ? $change_options{$unit}->{plural} : "${unit}s";
    }
    say "$number $unit";
    $change -= $number * $value;
  }
}

print <<'__END_OF_INTRO';
                              Change
               Creative Computing  Morristown, New Jersey


I, Your friendly microcomputer, will determine
the correct change for items costing up to $100.


__END_OF_INTRO

while ( 1 ) {
  my $cost = get_pennies( 'Cost of item' );
  my $payment = get_pennies( 'Amount of payment');

  my $change = $payment - $cost;
  my $change_formatted = sprintf( "%.2f", $change / 100 );
  if ( $change == 0 ) {
    say 'Correct amount, thank you.';
  } elsif ( $change < 0 ) {
    say 'Sorry, you have short-changed me $', abs($change_formatted);
  } else {
    say 'Your change, $', $change_formatted;
    make_change( $change );
    say "Thank you, come again\n\n";
  }
}

File: 22_Change/javascript/change.js
// CHANGE
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// Main program
async function main()
{
    print(tab(33) + "CHANGE\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("I, YOUR FRIENDLY MICROCOMPUTER, WILL DETERMINE\n");
    print("THE CORRECT CHANGE FOR ITEMS COSTING UP TO $100.\n");
    print("\n");
    print("\n");
    while (1) {
        print("COST OF ITEM");
        a = parseFloat(await input());
        print("AMOUNT OF PAYMENT");
        p = parseFloat(await input());
        c = p - a;
        m = c;
        if (c == 0) {
            print("CORRECT AMOUNT, THANK YOU.\n");
        } else {
            print("YOUR CHANGE, $" + c + "\n");
            d = Math.floor(c / 10);
            if (d)
                print(d + " TEN DOLLAR BILL(S)\n");
            c -= d * 10;
            e = Math.floor(c / 5);
            if (e)
                print(e + " FIVE DOLLAR BILL(S)\n");
            c -= e * 5;
            f = Math.floor(c);
            if (f)
                print(f + " ONE DOLLAR BILL(S)\n");
            c -= f;
            c *= 100;
            g = Math.floor(c / 50);
            if (g)
                print(g + " ONE HALF DOLLAR(S)\n");
            c -= g * 50;
            h = Math.floor(c / 25);
            if (h)
                print(h + " QUARTER(S)\n");
            c -= h * 25;
            i = Math.floor(c / 10);
            if (i)
                print(i + " DIME(S)\n");
            c -= i * 10;
            j = Math.floor(c / 5);
            if (j)
                print(j + " NICKEL(S)\n");
            c -= j * 5;
            k = Math.floor(c + 0.5);
            if (k)
                print(k + " PENNY(S)\n");
            print("THANK YOU, COME AGAIN.\n");
            print("\n");
            print("\n");
        }
    }
}

main();
